File: 42-Trapping-Rain-Water.cs
using System;
namespace AlgoPractice
{
	public class Solution
	{
		public int Trap(int[] height)
		{
			var left = 0;
			var right = height.Length - 1;

			var maxLeft = height[left];
			var maxRight = height[right];

			var result = 0;

			while(left < right)
            {
				if(maxLeft < maxRight)
                {
					left++;
					var currentResult = maxLeft - height[left];
					if(currentResult > 0)
					{
						result += currentResult;
					}
					maxLeft = Math.Max(maxLeft, height[left]);
                }
                else
                {
					right--;
					var currentResult = maxRight - height[right];
					if(currentResult > 0)
                    {
						result += currentResult;
                    }
					maxRight = Math.Max(maxRight, height[right]);
                }

            }
			return result;
		}
	}
}

File: kotlin/42-Trapping-Rain-Water.kt
import kotlin.math.max

class Solution {
    fun trap(height: IntArray): Int {
        var volumeOfWaterTrapped = 0
        var l = 0
        var r = height.lastIndex
        var currentLeftMax = -1
        var currentRightMax = -1

        while (l < r) {
            if (height[l] <= height[r]) {
                currentLeftMax = max(currentLeftMax, height[l])
                l++
                volumeOfWaterTrapped += (currentLeftMax - height[l]).coerceAtLeast(0)
            } else {
                currentRightMax = max(currentRightMax, height[r])
                r--
                volumeOfWaterTrapped += (currentRightMax - height[r]).coerceAtLeast(0)
            }
        }
        return volumeOfWaterTrapped
    }
}